src/thread_event.c

```diff
@@ -0,0 +1,277 @@
+thread_prof_sample_event_handle(tsd_t *tsd) {
+	assert(config_prof && opt_prof);
+	if (prof_sample_event_wait_get(tsd) == 0U &&
+	    !prof_active_get_unlocked()) {
+		/*
+		 * If prof_active is off, we reset prof_sample_event_wait to be
+		 * the sample interval when it drops to 0, so that there won't
+		 * be excessive routings to the slow path, and that when
+		 * prof_active is turned on later, the counting for sampling
+		 * can immediately resume as normal.
+		 */
+		thread_prof_sample_event_update(tsd,
+		    (uint64_t)(1 << lg_prof_sample));
+	}
+}
+
+void
+thread_event_trigger(tsd_t *tsd, uint64_t usize, bool delay_event) {
+	/* usize has already been added to thread_allocated. */
+	uint64_t thread_allocated_after = thread_allocated_get(tsd);
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t accumbytes = thread_allocated_after -
+	    thread_allocated_last_event_get(tsd);
+
+	if (accumbytes < usize) {
+		/* accumbytes overflowed. */
+		accumbytes = UINT64_MAX;
+	}
+
+	uint64_t wait = THREAD_EVENT_MAX_WAIT;
+	bool no_event_on = true;
+
+#define W(event, condition)						\
+	if (condition) {						\
+		no_event_on = false;					\
+		uint64_t event_wait = event##_event_wait_get(tsd);	\
+		thread_event_wait_assert(event_wait);			\
+		if (event_wait > accumbytes &&				\
+		    event_wait - THREAD_EVENT_MIN_WAIT >= accumbytes) {	\
+			event_wait -= accumbytes;			\
+		} else {						\
+			event_wait = 0;					\
+			if (delay_event || !tsd_nominal(tsd) ||		\
+			    tsd_reentrancy_level_get(tsd) > 0) {	\
+				event_wait = THREAD_EVENT_MIN_WAIT;	\
+			}						\
+		}							\
+		thread_event_wait_assert(event_wait);			\
+		event##_event_wait_set(tsd, event_wait);		\
+		/*							\
+		 * If there is a single event, then the remaining wait	\
+		 * time may become zero, and we rely on either the	\
+		 * event handle or a thread_event_update() call later	\
+		 * to properly set next_event; if there are multiple	\
+		 * events, then	here we can get the minimum remaining	\
+		 * wait time to	the next already set event.		\
+		 */							\
+		if (event_wait > 0 && event_wait < wait) {		\
+			wait = event_wait;				\
+		}							\
+	}
+
+	ITERATE_OVER_ALL_WAITS
+#undef W
+
+	/*
+	 * In case none of the events has its condition on, set next_event to
+	 * be THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX to minimize malloc() slow
+	 * path triggering.
+	 */
+	if (no_event_on) {
+		/* The subtraction is intentionally susceptible to underflow. */
+		wait = THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX -
+		    thread_allocated_after;
+		if (wait < THREAD_EVENT_MIN_WAIT) {
+			wait = THREAD_EVENT_MIN_WAIT;
+		}
+	}
+
+	thread_event_wait_assert(wait);
+	thread_allocated_last_event_set(tsd, thread_allocated_after);
+	thread_event_adjust_thresholds_helper(tsd, thread_allocated_after,
+	    wait, no_event_on);
+
+	thread_event_assert_invariants(tsd);
+
+#define W(event, condition)						\
+	if (condition) {						\
+		thread_##event##_event_handle(tsd);			\
```

Comment:
Only call handler if wait is 0.

Reply:
Also add `assert(!delay_event);` before calling the handler.

Reply:
Makes sense. Also added `tsd_nominal(tsd)` and `tsd_reentrancy_level_get(tsd) == 0` asserts.